lib/instrumentation/modules/http2.js

```diff
@@ -0,0 +1,69 @@
+'use strict'
+
+var debug = require('debug')('elastic-apm')
+var eos = require('end-of-stream')
+var shimmer = require('../shimmer')
+
+module.exports = function (http2, agent) {
+  debug('shimming http2.createServer function')
+  shimmer.wrap(http2, 'createServer', wrapCreateServer)
+  shimmer.wrap(http2, 'createSecureServer', wrapCreateServer)
+
+  return http2
+
+  // The `createServer` function will unpatch itself after patching
+  // the first server prototype it patches.
+  function wrapCreateServer (original) {
+    return function wrappedCreateServer (options, handler) {
+      var server = original.apply(this, arguments)
+      shimmer.wrap(server.constructor.prototype, 'emit', wrapEmit)
+      wrappedCreateServer.__obUnwrap()
+      return server
+    }
+  }
+
+  function wrapEmit (original) {
+    var patched = false
+    return function wrappedEmit (event, stream, headers) {
+      if (event === 'stream') {
+        if (!patched) {
+          patched = true
+          shimmer.wrap(stream.constructor.prototype, 'respond', wrapRespond)
+        }
+
+        // TODO: This is a pile of hacks to make HTTP/2 sockets look like
+        // HTTP/1 requests
+        stream.headers = headers
+        stream.socket = stream.session.socket
+        stream.method = headers[':method']
+
+        debug('intercepted request event call to http2.Server.prototype.emit')
+
+        var trans = agent.startTransaction()
+        trans.type = 'http2-request'
+        trans.req = stream
+        trans.res = stream
+
+        eos(stream, function (err) {
+          if (!err) return trans.end()
+
+          // Handle case where res.end is called after an error occurred on the
+          // stream (e.g. if the underlying socket was prematurely closed)
+          res.on('prefinish', function () {
+            trans.end()
+          })
+        })
+      }
+
+      return original.apply(this, arguments)
+    }
+  }
+
+  function wrapRespond (original) {
+    return function (headers) {
+      var trans = agent._instrumentation.currentTransaction
+      trans.result = 'HTTP ' + (headers[':status'] || '').toString()[0] + 'xx'
```

Comment:
Be careful that `trans` might be `null`

Reply:
Ah, yep. I'll fix that when I get more time to work on this.